File: SecureBlackbox Samples/PKCS11 Certificate Storage/c++/pkcs11certificatestorage.cpp
/*
 * SecureBlackbox 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of SecureBlackbox in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/secureblackbox
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sstream>
#include <string>
#include <iostream>
#include <vector>
#include "../../include/secureblackbox.h"

namespace ArgParser {
    static char* optval(int argc, char** argv, const char* option) {
        for (int x = 0; x < argc - 1; x++) {
            if (!strcmp(argv[x], option)) {
                return argv[x + 1];
            }
        }
        return (char*)"";
    }
};

using namespace std;
using namespace ArgParser;

void displayHelp(const char* errMes) {
    printf(
        "NAME\n"
        "  pkcs11certificatestorage -- SecureBlackbox CertificateStorage Demo Application\n\n"
        "SYNOPSIS\n"
        "  pkcs11certificatestorage <-storage driver_path> [-pin pin] [-slot slot_num]\n\n"
        "DESCRIPTION\n"
        "  This sample illustrates the use of CertificateStorage component to access HSMs via PKCS11 interface. \n\n"
        "  The options are as follows:\n\n"
        "  -storage      A path to the pkcs11 driver file (Required).\n\n"
        "  -pin          The user PIN for the device. If no PIN is provided, the sample won't be signing in.\n\n"
        "  -slot         The slot number to use. If not specified, the first slot with a token in will be used. Pass -1 as a slot number to only list the slots.\n\n"
        "EXAMPLES\n"
        "  pkcs11certificatestorage -storage C:\\pkcs11\\pkcs11.dll -pin mypassword\n\n"
    );

	if (strlen(errMes) > 0)
	{
		printf("Error: %s \n\n", errMes);
	}
}

int main(int argc, char** argv) {
    CertificateStorage cs;
	std::stringstream ss;
	auto slotList = std::vector<std::string>{};

    char storeuri[100];
    char confrequest[100];
    char parPin[32];
    char parSlot[32];
    char parLogin[32];

    int certidx = -1;

    // Validate input
    if (argc < 2) {
        displayHelp("");
        goto done;
    }

    char* parDrvPath = optval(argc, argv, "-storage");
    if (!strcmp(parDrvPath, "")) {
        displayHelp("-storage is required.");
        goto done;
    }

    char* pin = optval(argc, argv, "-pin");
    char* slot = optval(argc, argv, "-slot");

    // if slot is not provided, the first slot with the token in will be opened
    sprintf(parSlot, (strlen(slot) > 0) ? "&slot=%s" : "%s", slot);

    // if pin is provided, signing in with it
    sprintf(parPin, (strlen(pin) > 0) ? ":%s" : "%s", pin);
    sprintf(parLogin, "login=%s", (strlen(pin) > 0) ? "yes" : "no");

    // forming the storage URI
    sprintf(storeuri, "pkcs11://user%s@localhost/%s?%s%s", parPin, parDrvPath, parLogin, parSlot);

    // Opening the storage
    printf("Opening storage: %s\n\n", storeuri);
    if (cs.Open(storeuri)) {
        goto done;
    }

    printf("The storage has been opened. Browsing for slots...\n\n");

    // Listing slots
	ss << cs.ListStores();
	for (std::string line; std::getline(ss, line, '\n');)
		slotList.push_back(line);
	
	printf("%d slot(s) have been found\n\n", slotList.size());

	for (int x = 0; x < slotList.size(); x++) {
		sprintf(confrequest, "PKCS11SlotTokenPresent[%d]", x);
		const int tokenFound = !strcmp(cs.Config(confrequest), "True");

		if (tokenFound) {
			cout << "- Slot: " << (x + 1) << " (with token): " << slotList.at(x) << endl << endl;
		} 
		else {
			cout << "- Slot: " << (x + 1) << " (no token): " << slotList.at(x) << endl << endl;
		}
	}

    // Listing certificates
    printf("%d certificates(s) have been found (can be zero if an empty slot was selected)\n\n", cs.GetCertCount());

    for (int y = 0; y < cs.GetCertCount(); y++) {
        const int keyFound = cs.GetCertPrivateKeyExists(y);
        if ((keyFound) && (certidx < 0)) {
            certidx = y;
        }
        printf("- Cert %d: %s (%s)\n\n", y, cs.GetCertSubjectRDN(y), keyFound ? "private key found" : "no private key");
    }

    // Closing the storage
    cs.Close(false);

err:
    if (cs.GetLastErrorCode()) {
        printf("Error: [%i] %s\n", cs.GetLastErrorCode(), cs.GetLastError());
    }
done:
	printf("Press Enter to exit the demo.\n");
    getchar();
}

File: SecureBlackbox Samples/PKCS11 Certificate Storage/php/pkcs11_certificate_storage.php
<?php
/*
 * SecureBlackbox 2024 PHP Edition - Sample Project
 *
 * This sample project demonstrates the usage of SecureBlackbox in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/secureblackbox
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
require_once('../include/secureblackbox_pkcs11certificatestorage.php');
require_once('../include/secureblackbox_pdfsigner.php');
require_once('../include/secureblackbox_pdfverifier.php');
require_once('../include/secureblackbox_certificatestorage.php');
require_once('../include/secureblackbox_const.php');
?>
<style>
table { width: 100% !important; }
td { white-space: nowrap; }
td input { width: 100%; }
td:last-child { width: 100%; }
</style>

<div width="90%">
  <form method=POST>
    <h2>PDF Signing Demo</h2>
    
    <h3>General Options</h3>
    <table>
      <tr><td>Storage File:</td><td><input type=text name=storageFile value=""></td></tr>
      <tr><td>PIN:</td><td><input type=password name=sPIN value=""></td></tr>
    </table>
    <br/>
    <br/>

    <input type="submit" value="Open" />
  </form>
</div><br/>

<?php

  if ($_SERVER['REQUEST_METHOD'] == "POST") {
    $certstor = new SecureBlackBox_CertificateStorage();
    $certstor_dop = new SecureBlackBox_CertificateStorage();
    
    try
    {
      $certstor->doOpen('pkcs11:///' . $_REQUEST['storageFile'] . '?slot=-1');

      $slots = preg_split("/\r\n|\n|\r/", $certstor->doListStores());

      for ($i = 0; $i < count($slots); $i++)
      {
        $desc = $slots[$i];
        $active = $certstor->doConfig('PKCS11SlotTokenPresent[' . $i . ']');
        
        if ($desc != '')
        {
          if ($active == 'True')
          {
            
            echo $desc . ': <br/>';

            $certstor_dop->doOpen('pkcs11://user:' . $_REQUEST['sPIN'] . '@/' . $_REQUEST['storageFile'] . '?slot=' . $i);
            
            for ($j = 0; $j < $certstor_dop->getCertCount(); $j++)
            {
              echo '&nbsp;&nbsp;Subject: ' . $certstor_dop->getCertSubject($j) . '<br/>';
              echo '&nbsp;&nbsp;Issuer: ' . $certstor_dop->getCertIssuer($j) . '<br/>';
              echo '&nbsp;&nbsp;ValidFrom: ' . $certstor_dop->getCertValidFrom($j) . '<br/>';
              echo '&nbsp;&nbsp;ValidTo: ' . $certstor_dop->getCertValidTo($j) . '<br/>';
              echo '&nbsp;&nbsp;Key: ' . $certstor_dop->getCertKeyAlgorithm($j) . '  (' . $certstor_dop->getCertKeyBits($j) . ') <br/><br/>';
            }
          }
          else
          {
            echo $desc . ': No token <br/><br/>';
          }
        }
      }
    }
    catch (exception $e) {
      echo "<h2>Opening Failure (Details Below)</h2><p>" . $e->getMessage() . "</p>";
    }
  }
?>